core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -279,7 +279,7 @@ private object RTS {
         // (reverse chronological).
         stack.pop() match {
           case f0: Finalizer =>
-            val f: IO[Nothing, Option[List[Throwable]]] = f0.finalizer.run.map(collectDefect)
+            val f: IO[Nothing, Option[List[Throwable]]] = fork(f0.finalizer, _ => IO.unit).observe.map(collectDefect)
```

Comment:
I had another version that was much simpler:
```scala
fork(f0.finalizer, accumFailures)
```
which allowed me to get rid of `collectDefect` and `zipFailures`; however, the unhandled errors handler action is submitted on the thread pool, so there is no guarantee of ordering.

Reply:
Exactly, and finalizers must be linearly ordered.